burn-wgpu/src/kernel/matmul/tune.rs

```diff
@@ -0,0 +1,305 @@
+                    self.workgroup_size_x,
+                    self.workgroup_size_y
+                )
+            }
+        }
+    };
+}
+
+tiling2d_tunable!(
+    Tiling2DContiguousLoad,
+    kernel::matmul::contiguous::matmul_tiling_2d
+);
+
+tiling2d_tunable!(Tiling2DTileLoad, kernel::matmul::tile::matmul_tiling_2d);
+tiling2d_tunable!(
+    Tiling2DContiguousLoadVectorized,
+    kernel::matmul::contiguous_vectorized::matmul_tiling_2d
+);
+
+tiling2d_tunable!(
+    Tiling2DTileLoadVectorized,
+    kernel::matmul::tile_vectorized::matmul_tiling_2d
+);
+
+#[derive(new)]
+struct MemoryCoalescing<E: WgpuElement, const D: usize> {
+    workgroup_size_x: usize,
+    workgroup_size_y: usize,
+    _elem: PhantomData<E>,
+}
+
+impl<E: WgpuElement, const D: usize> KernelFunction for MemoryCoalescing<E, D> {
+    type Input = (WgpuTensor<E, D>, WgpuTensor<E, D>);
+    type Output = WgpuTensor<E, D>;
+
+    fn call(&self, (lhs, rhs): Self::Input) -> Self::Output {
+        matmul_mem_coalescing::<E, D>(lhs, rhs, self.workgroup_size_x, self.workgroup_size_y)
+    }
+
+    fn description(&self) -> String {
+        format!(
+            "Memory Coalescing matmul - W_X {}, W_Y {}",
+            self.workgroup_size_x, self.workgroup_size_y
+        )
+    }
+}
+
+#[derive(new)]
+struct MatmulBenchmark<F: WgpuElement, const D: usize> {
+    shape_lhs: Shape<D>,
+    shape_rhs: Shape<D>,
+    num_repeats: usize,
+    matmul: PhantomData<F>,
+    func: AutoTuneFunction<(WgpuTensor<F, D>, WgpuTensor<F, D>), WgpuTensor<F, D>>,
+}
+
+impl<E, const D: usize, G> Benchmark<G> for MatmulBenchmark<E, D>
+where
+    E: WgpuElement + FloatElement,
+    G: GraphicsApi,
+{
+    type Args = (WgpuTensor<E, D>, WgpuTensor<E, D>);
+
+    fn name(&self) -> String {
+        format!("{:?} x {:?}", self.shape_lhs.dims, self.shape_rhs.dims)
+    }
+
+    fn num_samples(&self) -> usize {
+        5
+    }
+
+    fn execute(&self, (lhs, rhs): Self::Args) {
+        for _ in 0..self.num_repeats {
+            self.func.call((lhs.clone(), rhs.clone()));
+        }
+    }
+
+    fn prepare(&self, device: &WgpuDevice) -> Self::Args {
+        let lhs = Tensor::<WgpuBackend<G, E, i32>, D>::random(
+            self.shape_lhs.clone(),
+            Distribution::Default,
+        )
+        .to_device(device);
+        let rhs = Tensor::<WgpuBackend<G, E, i32>, D>::random(
+            self.shape_rhs.clone(),
+            Distribution::Default,
+        )
+        .to_device(device);
+
+        (lhs.into_primitive(), rhs.into_primitive())
+    }
+}
+
+/// Choose the best matmul kernel by using autotuning.
+pub fn tune<G: GraphicsApi, E, const D: usize>(
+    lhs: WgpuTensor<E, D>,
+    rhs: WgpuTensor<E, D>,
+) -> WgpuTensor<E, D>
+where
+    E: WgpuElement + FloatElement,
+{
+    let (shape_lhs, shape_rhs) = calculate_benchmark_shapes(lhs.shape.clone(), rhs.shape.clone());
+    let id = AutoTuneKey::new(
+        vec![Vec::from(shape_lhs.dims), Vec::from(shape_rhs.dims)],
```

Comment:
I think there is a problem, we will recompute the benchmarks for tensors of different rank. we should only add M, N and K in the list of usize.